metabase/metabase
# Automated Releases from CI 🚀

Closes https://github.com/metabase/metabase-private/issues/168

### Description

This allows us to release v47 and later versions of metabase fully via github actions.  [Video Tour](https://www.loom.com/share/8f4319a6e39b4ccb9327c1d4d4c0b28e)

![Screen Shot 2023-09-19 at 8 30 27 AM](https://github.com/metabase/metabase/assets/30528226/544649b9-b539-43c5-834e-1ea2f0b9f861)



![Screen Shot 2023-09-18 at 1 42 41 PM](https://github.com/metabase/metabase/assets/30528226/82928ab6-332f-4dd4-bbeb-10b31fa277b3)
![Screen Shot 2023-09-18 at 1 42 55 PM](https://github.com/metabase/metabase/assets/30528226/09f0f743-2bcf-4387-bad7-a7bf54c801f7)
![Screen Shot 2023-09-18 at 1 43 07 PM](https://github.com/metabase/metabase/assets/30528226/95ab3b45-c9e8-4ed9-a45a-51f6483f41c7)

- **Build Step**
  - Uses unit-tested javascript version helpers to check generate release version numbers
  - Adds commit hash to artifact name to make fetching artifacts built from that commit simpler
- **Test Step**
  - simplifies artifact fetching code to match github docs
- **Release Step**
  - Is now triggered manually, giving us the freedom to release whenever we want from already-built artifacts
  - Adds s3 upload
  - Adds git tagging
  - Adds release notes generation
  - Adds version-info generation
  - Adds cloudfront invalidation

Example Run:
[Release 1: Build](https://github.com/iethree/metabase/actions/runs/6225770403) (17 min)
[Release 2: Test](https://github.com/iethree/metabase/actions/runs/6225153780) (34 min)
[Release 3: Publish](https://github.com/iethree/metabase/actions/runs/6226847160) (12 min)

We can have fully-tested oss and enterprise releases from any given commit in a release branch in 63 minutes.

If necessary, we can bypass the testing step and have published releases in 29 minutes.

In addition, these workflows are now (almost) fully parameterized via environment variables, so we can release from any fork of metabase that has variables properly set up. This allows us to dry-run our releases, as well as build security updates in CI from private repositories.

We will need to add a few new (non-secret) repository variables to enable this:
```
AWS_S3_STATIC_BUCKET
AWS_S3_DOWNLOADS_BUCKET
AWS_CLOUDFRONT_STATIC_ID
AWS_CLOUDFRONT_DOWNLOADS_ID
AWS_REGION

DOCKERHUB_REPO
DOCKERHUB_OWNER
DOCKERHUB_STAGING_REPO
```


### How to verify

Here are some tested scenarios:

- Happy Paths
  - latest: v0.47.30 ([build](https://github.com/iethree/metabase/actions/runs/6240210970)) ([test](https://github.com/iethree/metabase/actions/runs/6240417016)) ([publish](https://github.com/iethree/metabase/actions/runs/6241168487)) ([s3 OSS JAR](http://iethree-downloads.metabase.com.s3.us-east-1.amazonaws.com/v0.47.30/metabase.jar)) ([s3 EE JAR](http://iethree-downloads.metabase.com.s3.us-east-1.amazonaws.com/enterprise/v1.47.30/metabase.jar))
  - non-latest: 0.47.29([build](https://github.com/iethree/metabase/actions/runs/6240509796)) ([test](https://github.com/iethree/metabase/actions/runs/6240665836)) ([publish](https://github.com/iethree/metabase/actions/runs/6240771142)) ([s3 OSS JAR](http://iethree-downloads.metabase.com.s3.us-east-1.amazonaws.com/v0.47.29/metabase.jar)) ([s3 EE JAR](http://iethree-downloads.metabase.com.s3.us-east-1.amazonaws.com/enterprise/v1.47.29/metabase.jar))
  - release candidate: v0.47-rc9 ([build](https://github.com/iethree/metabase/actions/runs/6240489764)) ([test](https://github.com/iethree/metabase/actions/runs/6240636561)) ([publish](https://github.com/iethree/metabase/actions/runs/6240774452)) ([s3 OSS JAR](http://iethree-downloads.metabase.com.s3.us-east-1.amazonaws.com/v0.47-rc9/metabase.jar)) ([s3 EE JAR](http://iethree-downloads.metabase.com.s3.us-east-1.amazonaws.com/enterprise/v1.47-rc9/metabase.jar))
- Sad Paths 
  - [invalid build version string](https://github.com/iethree/metabase/actions/runs/6240583080): v9.47
  - [invalid build commit string](https://github.com/iethree/metabase/actions/runs/6240587271/job/16941011895): v0.47.23 (put the version in the commit field)
  - [release version string](https://github.com/iethree/metabase/actions/runs/6240453424/job/16940588533): v9.47
  - [release commit string](https://github.com/iethree/metabase/actions/runs/6240469360)
  - [valid release commit that was never built](https://github.com/iethree/metabase/actions/runs/6240435704)

[Github Releases](https://github.com/iethree/metabase/releases) | [OSS Dockerhub](https://hub.docker.com/r/iethree/metabase/tags) | [Enterprise Dockerhub](https://hub.docker.com/r/iethree/metabase-enterprise/tags) | [Staging Dockerhub](https://hub.docker.com/r/iethree/mb-staging/tags) | [version-info](http://iethree-static.metabase.com.s3.us-east-1.amazonaws.com/version-info.json) | [version-info-ee](http://iethree-static.metabase.com.s3.us-east-1.amazonaws.com/version-info-ee.json)

A couple aspects of this have not been tested
- cloudfront invalidations
- docs repo trigger
- EBS artifact generation and push

These are prohibitively difficult to test in a fork, and should be pretty safe, also, they have very easy manual fallbacks (aws cli, manual docs repo update, running the EBS artifact generation locally)

## Still Todo

- Allow releases without github actions (very nearly complete)
- Allow releases without github
- Allow automated CI releases for older metabase versions
- Remove unused clojure code in `src/release`

### Checklist
